### PR TITLE
HttpServer

### PR DESCRIPTION
**Pull request by sagizagagi**

### CODE REVIEW

File: CSVSource/build.gradle [hunk 1]
```diff
@@ -35,6 +35,12 @@ applicationDistribution.from("src/main/resources") {
 }
 
 dependencies {
+    // https://mvnrepository.com/artifact/com.sun.net.httpserver/http
+    //implementation group: 'com.sun.net.httpserver', name: 'http', version: '20070405'
+     compile 'com.sun.net.httpserver:http:20070405'
```
Comment (fhcarter) [inline CSVSource/build.gradle:40]: What is the licensing on this?  Many of the Oracle/Sun things have a license that we generally don't like.  While this is somewhat separate, we'll need to document the licensing here for customers who use this to accept.

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 2]
```diff
@@ -46,6 +46,8 @@
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Stream;
 
+import javax.swing.plaf.metal.OceanTheme;
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:49]: if we don't use the import, please remove it.

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 3]
```diff
@@ -145,6 +147,8 @@
     ExecutorService executionPool = null;
     ExtensionWebSocketClient oClient;
 
+    XMLHttpServer xmlHttpServer ;
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:150]: Yup

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 4]
```diff
@@ -156,6 +160,12 @@
     boolean deleteAfterProcessing = false;
     int pollTime;
 
+
+    Boolean enableHttpListener = false ;
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:164]: yup. Please fix these up

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 5]
```diff
@@ -156,6 +160,12 @@
     boolean deleteAfterProcessing = false;
     int pollTime;
 
+
+    Boolean enableHttpListener = false ; 
+    int port ; 
+    String context; 
+    String IPListenAddress;
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:167]: This is important -- only classes start with upper case

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 6]
```diff
@@ -264,6 +279,27 @@ void prepareConfigurationData() {
         executionPool = new ThreadPoolExecutor(maxActiveTasks, maxActiveTasks, 0l, TimeUnit.MILLISECONDS,
                 new LinkedBlockingQueue<Runnable>(maxQueuedTasks));
 
+        if (options.get(ENABLE_HTTP_LISTNER) != null){
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:282]: please

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 7]
```diff
@@ -264,6 +279,27 @@ void prepareConfigurationData() {
...
+            enableHttpListener = (Boolean) options.get(ENABLE_HTTP_LISTNER);
+        }
+
+        if (enableHttpListener){
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:286]: please

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 8]
```diff
@@ -264,6 +279,27 @@ void prepareConfigurationData() {
...
+            port = 8001; 
+            if (options.get(PORT) != null){
+                port = (int) options.get(PORT); 
+            }
+            context = "/alarm";
+            if (options.get(HTTP_CONTEXT) != null){
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:292]: all of these

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 9]
```diff
@@ -264,6 +279,27 @@ void prepareConfigurationData() {
...
+                context = (String) options.get(HTTP_CONTEXT); 
+            }
+            IPListenAddress = "localhost";
+            if (options.get(LISTEN_ADDRESS) != null){
+                IPListenAddress = (String) options.get(LISTEN_ADDRESS); 
+            }
+
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:299]: don't need this empty space

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 10]
```diff
@@ -264,6 +279,27 @@ void prepareConfigurationData() {
...
+
+        }
+
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:302]: or this

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 11]
```diff
@@ -310,6 +346,20 @@ public void run() {
...
+                if (xmlHttpServer==null){
+                    xmlHttpServer = new XMLHttpServer();
+                    xmlHttpServer.oClient = oClient; 
+                    xmlHttpServer.port = port; 
+                    xmlHttpServer.context1 = context; 
+                    xmlHttpServer.IPListenAddress = IPListenAddress;
+                    xmlHttpServer.Start();
+                }
+            }
+
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:361]: whitespace

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+public class XMLHttpServer {
+    Logger log = LoggerFactory.getLogger(this.getClass().getCanonicalName());
+
+    HttpServer server;
+    ThreadPoolExecutor threadPoolExecutor;
+    public ExtensionWebSocketClient oClient = null; 
+    public String context1 = "/alarm";
+    public int port = 8001;
+    public String IPListenAddress = "localhost";
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java:35]: `ipListenAddress` please

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+    public Boolean Stop(){
+        
+        if (server != null){
+            server.stop(0);
+        }
+        return true; 
+    }
+
+    public Boolean Start() {
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java:45]: methods start with lower case

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+        return true;
+    }
+
+    private class MyHttpHandler implements HttpHandler {
+
+        @Override
+
+        public void handle(HttpExchange httpExchange) throws IOException {
+
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java:73]: don't need to double space this whole thing

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+            String requestParamValue = null;
+
+            String method = httpExchange.getRequestMethod();
+
+            if ("POST".equals(method)) {
+
+                log.info("Accepting POST HTTP Request");
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java:80]: These should probably be at `log.debug` level -- they are just for debugging and we'll want to turn them off.

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+                String line = null; 
+                String value = "";
+                while ((line = br.readLine()) != null ){
+                    value += line; 
+                }
+                log.error(value);
+
+                try {
+                    Boolean extendedLogging = true; 
+                    JSONObject json = XML.toJSONObject(value); // converts xml to json
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java:94]: seems like we should be checking the information format sent.  Cannot just assume it's XML?

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+                    String jsonPrettyPrintString = json.toString(); // json pretty print
+                    if (extendedLogging){
+                        log.info(jsonPrettyPrintString);
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java:97]: `.debug` better here, too

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+                    Map<String, String> lineValues = new HashMap<String, String>();
+                    lineValues.put("xml", jsonPrettyPrintString);
+                    file.add(lineValues);
+    
+                    CSVReader.sendNotification("TCP","XML", 0, file, oClient);
+                } catch (JSONException je) {
+                    log.error("Convert2Json failed", je);
+                    requestParamValue = je.toString(); 
+                }
+
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java:110]: spaces

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+import org.json.JSONException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vantiq.extjsdk.ExtensionWebSocketClient;
+
+import com.sun.net.httpserver.*;
+
+public class XMLHttpServer {
```
Comment (fhcarter) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/XMLHttpServer.java:27]: Need some description here of what's going on.

**Review by fhcarter — APPROVED** [review]
Please fix up slider comments.  I didn't see any that were wrong.

Mostly OK -- spacing, etc., and style, but sider covered most of those.

Licensing issues about sun's http server.  May be better choices.

And you don't need approval to merge from one branch to another -- only to master.

File: CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java [hunk 20]
```diff
@@ -144,6 +134,7 @@
     // Components used
     ExecutorService executionPool = null;
     ExtensionWebSocketClient oClient;
+    XMLHttpServer xmlHttpServer;
```
Comment (sagizagagi) [inline CSVSource/src/main/java/io/vantiq/extsrc/CSVSource/CSV.java:137]: > __[PMD Java]__ A high number of imports can indicate a high degree of coupling within an object. ([view](https://sider.review/gh/repos/141492069/pulls/254#issue-5581421))
>
> <details>
>
> | Rule | Ruleset | Priority |
> | ---- | ------- | -------- |
> | `ExcessiveImports` | Design | 3 |
>
> References:
> - <https://pmd.github.io/pmd-6.33.0/pmd_rules_java_design.html#excessiveimports>
>
> You can [close this issue](https://sider.review/gh/repos/141492069/pulls/254#issue-5581421) if no need to fix it. [Learn more](https://help.sider.review/getting-started/working-with-issues).
>
> </details>

.

**Comment by sagizagagi** [comment]
Ability to receive XML object from post request, convert to JSON and notify Vantiq

**Pull request merged**